File: Breadth-First Search/Topological sort/LET_no886_Possible Bipartition.cpp
class Solution {
public:
    bool possibleBipartition(int n, vector<vector<int>>& dislikes) {
        unordered_map<int, vector<int>> graph;
        vector<int> colors(n + 1, -1);
        for (auto pair : dislikes) {
            int a = pair[0], b = pair[1];
            graph[a].push_back(b);
            graph[b].push_back(a);
        }

        for (int i = 1; i <= n; i++) {
            if (colors[i] == -1 && !bfs(colors, graph, i))
                return false;
        }

        return true;   
    }

    bool bfs(vector<int>& colors, unordered_map<int, vector<int>>& graph, int start) {
        queue<int> q;
        q.push(start);
        colors[start] = 0;

        while (!q.empty()) {
            int n = q.size();
            while (n > 0) {
                int cur = q.front();
                q.pop();
                for (int nxt : graph[cur]) {
                    if (colors[nxt] == colors[cur])
                        return false;
                    else if (colors[nxt] == -1) {
                        colors[nxt] = 1 - colors[cur];
                        q.push(nxt);
                    }
                }
                n--;
            }
        }

        return true;
    }
};

File: Breadth-First Search/Topological sort/LET_no886_Possible Bipartition.py
class Solution:
    def possibleBipartition(self, n: int, dislikes: List[List[int]]) -> bool:
        dislike_graph = collections.defaultdict(list)
        for a, b in dislikes:
            dislike_graph[a].append(b)
            dislike_graph[b].append(a)

        colors = [-1] * (n + 1)

        for i in range(1, n + 1):
            if colors[i] == -1 and not self.bfs(colors, dislike_graph, i):
                return False

        return True
    
    def bfs(self, colors, graph, start):
        queue = collections.deque([start])
        colors[start] = 0
        while queue:
            for _ in range(len(queue)):
                cur = queue.popleft()
                for nxt in graph[cur]:
                    if colors[nxt] == -1:
                        colors[nxt] = 1 - colors[cur]
                        queue.append(nxt)
                    elif colors[nxt] == colors[cur]:
                        return False
            
        return True